### ISSUE TITLE
Rate limit exceeded during pull-request

### ISSUE THREAD

**Issue by herbaltealeaf** [BetaHuhn/repo-file-sync-action#126]
**🐞 Describe the bug**

Appears that when attempting to sync many repos github may throttle the number of pull-requests that can be made and will result in only partially completing the full list.

```
...
Locally syncing file(s) between source and target repository
Source is directory
Pushing changes to target repository
Creating new PR
Error: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Error: HttpError: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Repository Info
Slug		: client-config
Owner		: acme-corp
Https Url	: https://github.com/acme-corp/client-config
Branch		: master
	
Locally syncing file(s) between source and target repository
Source is directory
Pushing changes to target repository
Creating new PR
Error: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Error: HttpError: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Repository Info
Slug		: app-config
Owner		: acme-corp
Https Url	: https://github.com/acme-corp/app-config
Branch		: master
	
Locally syncing file(s) between source and target repository
Source is directory
Pushing changes to target repository
Creating new PR
Error: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Error: HttpError: You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
Repository Info
Slug		: api-config
Owner		: acme-corp
Https Url	: https://github.com/acme-corp/api-config
Branch		: master
...
...
```

**📚 To Reproduce**

Perform many syncs within one of the limits. Could be x per hour or x per minute.

**💡 Expected behavior**

Either fast-fail and exit non-zero or retry after 5-10 seconds as per some (_additional_) action input. The action shouldn't indicate that it was successful if it didn't actually perform all the updates in the config (`sync.yml`).

**🖼️ Screenshots**

![image](https://user-images.githubusercontent.com/41610038/138177531-8b8e1be9-d46a-48ca-8061-d8a6bb67914f.png)

_notice the matrix shows 12 jobs, but my `sync.yml` has 20 repos listed._

![image](https://user-images.githubusercontent.com/41610038/138176539-37855ac8-5f0f-4c33-b0ab-cfbe88b87355.png)

**⚙️ Environment**
- Action version: BetaHuhn/repo-file-sync-action@v1.15.0

**📋 Additional context**

**Comment by samschurter** [comment]
Plus one for this, I just added the action and configured it to sync 12 repos. It succeeded on 10 and failed on the last 2.

![image](https://user-images.githubusercontent.com/7039470/139929076-9b89fbf1-70e9-4589-86a4-3f139dda836d.png)

**Comment by raeganbarker** [comment]
I've just started encountering the same issue as well. The rate limit error seems to occur after the 10th repo in a sync.

**Comment by BetaHuhn** [comment]
Just for clarity, did it work before with >10 repos and only now you encounter the rate limit @raeganbarker?

> p.s. Love this action. It's saved me so much time and manual effort!

Thanks! Glad you like it :)

**Comment by cansavvy** [comment]
I'm also encountering this error and I have an interest in getting around it! I think mine worked on 7 repositories before failing. But also I think my trigger may have been too lenient, the sync got triggered 3 times it seemed? https://github.com/jhudsl/DaSL_Course_Template_Bookdown/actions

I don't know JS but I'm happy to help anyway I can.

**Comment by BetaHuhn** [comment]
I still think the problem is that [`plugin-throttling.js`](https://github.com/octokit/plugin-throttling.js) is not being triggered. I will have to see how I can verify this.

The last change (https://github.com/BetaHuhn/repo-file-sync-action/commit/3007ff6ce457682c67154078f0a0e86892d9423b) to the plugin was a version bump with https://github.com/BetaHuhn/repo-file-sync-action/pull/105.

I will try to debug if that version changed something which could be a cause for the plugin to not trigger.

**Comment by BetaHuhn** [comment]
Looking at the code where the plugin is used, it should log a warning when hitting the limit:

https://github.com/BetaHuhn/repo-file-sync-action/blob/57f21592446fdcf96d38ceca9506b471670cfe5b/src/git.js#L31

I can't see that being logged in any of your logs.

This means the plugin is either not triggered because it fails somewhere or the rate limit is hit outside of the plugins control i.e. the octokit instance.

While looking around for similar issues I discovered https://github.com/peter-evans/create-pull-request/issues/855#issuecomment-852565210, this might be our issue as well. The plugin can't control the rate of `git push` requests, maybe they are hitting a rate limit at GitHub.

**Comment by BetaHuhn** [comment]
Okay I can now reproduce this issue: https://github.com/BetaHuhn/gh-repo-automation/runs/4096146204?check_suite_focus=true#step:3:326

**Comment by cansavvy** [comment]
> I can't see that being logged in any of your logs.

If you are looking at the most recent actions, I returned to an earlier version of `BetaHuhn/repo-file-sync-action` so that I could get this release out -- the earlier version, `BetaHuhn/repo-file-sync-action@v1.7.1` which works (it is after #49 is incorporated, so that's why I chose that one). 

This one shows the message that others have noted on this thread. 
https://github.com/jhudsl/DaSL_Course_Template_Bookdown/actions/runs/1417570063

**Comment by BetaHuhn** [comment]
> So the fix you implemented here: cee8520 worked, so if there's a way for you to incorporate it back in to the current version.

That's the thing, it is still there and wasn't changed since it was added. The only change was that the plugin used was bumped from 3.5.1 to 3.5.2.

I will revert that change and see if it works after that.

**Comment by cansavvy** [comment]
Is it as simple as updating this part?: https://github.com/BetaHuhn/repo-file-sync-action/blob/8c50c0d781d6ef22a5b395616897adda0cb6d0b8/dist/index.js#L17008

Looks like octokit/plugin-throttling.js only changes are just having to do with changing `abuse` -> `secondary rate` See: https://github.com/octokit/plugin-throttling.js/commit/6c361d45f07b8adf5fda11340ffaf3aee8a1ecfb 

I don't know any JS, just looking at where these things are referenced.

**Comment by BetaHuhn** [comment]
They changed `abuse` to `secondary rate` internally as GitHub changed the response, but the `onAbuseLimit` method stayed the same.

I tried a few older versions of this action and [v1.8.0](https://github.com/BetaHuhn/repo-file-sync-action/releases/tag/v1.8.0) also fails:

https://github.com/BetaHuhn/gh-repo-automation/runs/4096456566?check_suite_focus=true#step:3:381

You mentioned [v1.7.1](https://github.com/BetaHuhn/repo-file-sync-action/releases/tag/v1.7.1) works, so the issue must have been introduced with v1.8.0.

To verify this, could you repeat the same run with v1.7.1 and v1.8.0 (only changing the version) @cansavvy ?

**Comment by cansavvy** [comment]
Yes, v1.8.0 works too: https://github.com/jhudsl/DaSL_Course_Template_Bookdown/runs/4096640405?check_suite_focus=true

**Comment by BetaHuhn** [comment]
Then I am completely puzzled as v1.8.0 fails for me 😭

**Comment by herbaltealeaf** [comment]
> Yes, v1.8.0 works too: https://github.com/jhudsl/DaSL_Course_Template_Bookdown/runs/4096640405?check_suite_focus=true

Just to clarify, It doesn't fail on "no changes" and "overwriting pr" as this doesn't invoke the "new pr" api. Try removing all existing sync prs, making some minor (non-impactful) change, and then init the sync.

**Comment by cansavvy** [comment]
> Just to clarify, It doesn't fail on "no changes" and "overwriting pr" as this doesn't invoke the "new pr" api. Try removing all existing sync prs, make some minor (non-impactful) change to the source files, and then init the sync.

OH I see. Well then, unfortunately I won't be able to test this for you just yet then. Once those PRs are taken care of by the respective owners of those repos, I can try again later. Sorry.

**Comment by BetaHuhn** [comment]
Ah yeah, thanks @herbaltealeaf, I missed that when looking through the logs. That explains why it worked for @cansavvy but not for me.

**Comment by BetaHuhn** [comment]
Finally had some time today to further debug this issue. **Good news**: I found a solution which appears to work!

In the end it was quite simple, I just needed to install `@octokit/core` as dev dependency for [`plugin-throttling.js`](https://github.com/octokit/plugin-throttling.js) to work properly.

The fix is live on the [fix/rate-limit](https://github.com/BetaHuhn/repo-file-sync-action/tree/fix/rate-limit) branch and can be used like this:

```yml
uses: BetaHuhn/repo-file-sync-action@fix/rate-limit
```

If the action hits the rate limit it will log a warning and try again after a few seconds.

Please let me know if it works and I will merge this into master next week.

**Comment by raeganbarker** [comment]
This looks good on my side! I re-tested with a sync to 11 repos. The sync for the 10th repo initially ran into the GitHub API rate limit and then was handled successfully with the wait/retry logic.

**Comment by raeganbarker** [comment]
I noticed one thing. In the Job Summary Annotations section, the annotation doesn't indicate the repo sync succeeded or include the URL to the PR for the retried repo sync.
`Hit secondary GitHub API rate limit, retrying after 5s`
![image](https://user-images.githubusercontent.com/59625126/140800632-2f5cb974-9f04-4c0e-a588-386bd6a04772.png)
All the other annotations include the repo name + link to the PR created.

**Comment by BetaHuhn** [comment]
Thanks for testing it @raeganbarker!

I will take a look at the annotations issue you mentioned and release the fix tomorrow 👍

**Comment by BetaHuhn** [comment]
I just verified your issue and there seems to be a limit to 10 notices per action run being shown in the GitHub UI, that's why some of the syncs didn't show up for you @raeganbarker. Not much we can do about that.

I changed the rate limit warning to a debug log as I don't think it needs to create a notice each time the rate limit is hit and the action retries. This will happen quite often and doesn't need to be shown as a warning as it is the intended behavior.

**Comment by BetaHuhn** [comment]
This issue has been resolved in version [1.15.2](https://github.com/BetaHuhn/repo-file-sync-action/releases/tag/v1.15.2) :tada:

To use the latest version, use the Action with the `v1` tag:

```yml
uses: BetaHuhn/repo-file-sync-action@v1
```

Or the `latest` tag:

```yml
uses: BetaHuhn/repo-file-sync-action@latest
```

---

Sorry this took so long to fix! If you depend on this action or use it in a commercial project, maybe *consider* [sponsoring](https://github.com/sponsors/BetaHuhn) me on GitHub or [supporting](https://github.com/BetaHuhn/repo-file-sync-action#-about) me via platforms like [Ko-Fi](https://ko-fi.com/betahuhn). This keeps me motivated and allows me to spend more time on projects like this, which ultimately benefits you as a user of this action. Thanks!
